src/transformations/horner.cpp

```diff
@@ -244,29 +246,93 @@ summing the tiny high order elements first.
     range =  transformation->range;
 
+    const bool iterative_inverse = direction == PJ_INV && transformation->has_only_fwd;
 
     if (direction==PJ_FWD) {                              /* forward */
-        tcx = transformation->fwd_u + sz;
-        tcy = transformation->fwd_v + sz;
         e   = position.u - transformation->fwd_origin->u;
         n   = position.v - transformation->fwd_origin->v;
     } else {                                              /* inverse */
-        tcx = transformation->inv_u + sz;
-        tcy = transformation->inv_v + sz;
-        e   = position.u - transformation->inv_origin->u;
-        n   = position.v - transformation->inv_origin->v;
+        if (!iterative_inverse) {
+            e   = position.u - transformation->inv_origin->u;
+            n   = position.v - transformation->inv_origin->v;
+        } else {
+            // in this case fwd_origin needs to be added in the end
+            e = position.u;
+            n = position.v;
+        }
     }
 
     if ((fabs(n) > range) || (fabs(e) > range)) {
         proj_errno_set(P, PROJ_ERR_COORD_TRANSFM_OUTSIDE_PROJECTION_DOMAIN);
         return uv_error;
+    } else if (iterative_inverse) {
+        /*
+         * solve iteratively
+         *
+         * | E |   | u00 |   | u01 + u02*x + ...         ' u10 + u11*x + u20*y + ... | | x |
+         * |   | = |     | + |-------------------------- ' --------------------------| |   |
+         * | N |   | v00 |   | v10 + v11*y + v20*x + ... ' v01 + v02*y + ...         | | y |
+         *
+         * | x |   | Ma ' Mb |-1 | E-u00 |
+         * |   | = |-------- |   |       |
+         * | y |   | Mc ' Md |   | N-v00 |
+         */
+        const int order = transformation->order;
+        const double tol = transformation->inverse_tolerance;
+        const double de = e - transformation->fwd_u[0];
+        const double dn = n - transformation->fwd_v[0];
+        double x0 = 0.0;
+        double y0 = 0.0;
+        int loops = 32; // usually converges really fast (1-2 loops)
+        bool converged = false;
+        while (loops-- > 0 && !converged) {
+            double Ma = 0.0;
+            double Mb = 0.0;
+            double Mc = 0.0;
+            double Md = 0.0;
+            /* Coefficient pointers */
+            double *tcx = transformation->fwd_u;
+            double *tcy = transformation->fwd_v;
+            for (int i = 0; i <= order; ++i) {
+                for (int j = 0; j <= (order-i); ++j) {
+                    if (i == 0 && j == 0) {
+                        // do nothing
+                    } else if (i == 0) {
+                        Ma += (*tcx) * pow(x0, j-1);
+                        Md += (*tcy) * pow(y0, j-1);
+                    } else {
+                        Mb += (*tcx) * pow(y0, i-1) * pow(x0, j);
+                        Mc += (*tcy) * pow(x0, i-1) * pow(y0, j);
+                    }
+                    tcx++;
+                    tcy++;
+                }
```

Comment:
This seems a bit odd: The overall name of the procedure is "Horner" for a reason - because we use Horner's scheme for the polynomium evaluations. We do that for two other reasons:

1. Because of the stability of Horner's scheme due to its starting the summation with the smallest (i.e. high order) terms,
2. Because of its speed. On many platforms, the `pow(...)` call is **extremely** slow.

You probably should refactor the block labelled `The melody of this block is straight out of the great Engsager/Poder songbook` below into a separate function, and call that where needed.

Reply:
@busstoptaktik thank you for the detailed review! I was waiting for this comment, i will of course make the optimization. I did this already with the complex case, but i left this in this form to be a little more understandable for the time being.

Reply:
Awesome!